### PR TITLE
Add ability to rotate robots in Standalone Simulator GUI. 

### LINKED ISSUE #1539: SimulatorGUI: Set robot orientation
### Description of the task

as titled

Being able to set the robot orientation from the GUI is useful to set up specific scenarios to test various AI / robot functionalities.

There are many ways this could be done from a UI/UX perspective so here are some options:
* Click the robot, and then rotate it to face the cursor. It's orientation will be locked in when the mouse is clicked again
* Pop up a text-box and input the orientation manually

### Acceptance criteria

- [ ] as titled

### PR DESCRIPTION
**Pull request by PhilD71**
### Description
The ability to rotate robots has been added. This can be done by pressing the R key, and then moving the mouse around. The robot will stay in place, and rotate to face the cursor. Upon release of R, the robot is locked at that orientation.

### Testing Done
Tested in GUI. Made sure that pressing R while not on a robot does nothing, also made sure no other modifier combines with R to cause issues. 

### Resolved Issues
resolves #1539 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/software/gui/standalone_simulator/widgets/standalone_simulator_draw_function_visualizer.cpp [hunk 1]
```diff
@@ -21,6 +21,22 @@ void StandaloneSimulatorDrawFunctionVisualizer::setStandaloneSimulator(
...
+    {
+        r_clicked = true;
+    }
+}
+
+void StandaloneSimulatorDrawFunctionVisualizer::keyReleaseEvent(QKeyEvent* event)
+{
+    if (event->key() == Qt::Key_R)
+    {
+        r_clicked = false;
```
Comment (jonathanlew) [inline src/software/gui/standalone_simulator/widgets/standalone_simulator_draw_function_visualizer.cpp]: I don't think we need to stop rotating once the R key is released. It's find to stop the rotation as soon as the mouse is released

File: src/software/gui/standalone_simulator/widgets/standalone_simulator_draw_function_visualizer.h [hunk 2]
```diff
@@ -45,14 +47,26 @@ class StandaloneSimulatorDrawFunctionVisualizer : public DrawFunctionVisualizer
...
 
     // The final point when adding force to the ball, if any
     Point final_point;
 
+    // The initial point when orienting robot
+    Point rob_initial_point;
+
+    // The final point with orienting robot
+    Point rob_final_point;
+
```
Comment (jonathanlew) [inline src/software/gui/standalone_simulator/widgets/standalone_simulator_draw_function_visualizer.h]: why can't we use initial_point/final_point here?

File: src/software/simulation/physics/physics_robot.cpp [hunk 3]
```diff
@@ -371,6 +371,25 @@ void PhysicsRobot::setPosition(const Point& position)
     post_physics_step_functions.emplace(func);
 }
 
+void PhysicsRobot::setPositionOrientation(const Point& position, const Angle& angle)
```
Comment (jonathanlew) [inline src/software/simulation/physics/physics_robot.cpp]: Couldn't we just do setOrientation here?

**Review by jonathanlew — CHANGES_REQUESTED** [review]
works pretty well locally, just a few comments

File: src/software/gui/standalone_simulator/widgets/standalone_simulator_draw_function_visualizer.h (hunk 2 above)
Comment (PhilD71) [reply src/software/gui/standalone_simulator/widgets/standalone_simulator_draw_function_visualizer.h]: The reason I used separate fields is that using initial and final points creates a line on the screen, as if the robot was the ball. 
Screenshot from 2021-03-12 15-55-04

File: src/software/simulation/physics/physics_robot.cpp (hunk 3 above)
Comment (PhilD71) [reply src/software/simulation/physics/physics_robot.cpp]: Good call, I'll make that change! `setPosition` only takes a position parameter, though, so I'll modify it, and anyone that uses it to take two parameters, keeping the angle the same if there's no angle change.

File: src/software/gui/standalone_simulator/widgets/standalone_simulator_draw_function_visualizer.cpp (hunk 1 above)
Comment (PhilD71) [reply src/software/gui/standalone_simulator/widgets/standalone_simulator_draw_function_visualizer.cpp]: Both shift click and ctrl click work this way, so I'll change it to be consistent

File: src/software/gui/standalone_simulator/widgets/standalone_simulator_draw_function_visualizer.h (hunk 2 above)
Comment (jonathanlew) [reply src/software/gui/standalone_simulator/widgets/standalone_simulator_draw_function_visualizer.h]: yikes we need to fix that

Comment (PhilD71) [reply src/software/gui/standalone_simulator/widgets/standalone_simulator_draw_function_visualizer.h]: I think the problem might be a little worse than I thought, I tried to revert the change and now it just draws a line on the screen, doesn't even rotate

Comment (jonathanlew) [reply src/software/gui/standalone_simulator/widgets/standalone_simulator_draw_function_visualizer.h]: `getDrawBallVelocityFunction` can be detached from `initial_point` and `final_point`if we store the current ball velocity draw function as a private member and only initialize it if we're control clicking a ball. Then, `getDrawBallVelocityFunction` can just return this private member.

File: src/software/simulation/physics/physics_robot.h [hunk 4]
```diff
@@ -174,13 +174,13 @@ class PhysicsRobot
...
-     * will maintain its orientation, but will have its linear and angular
-     * velocity set to zero.
+     * Sets the position and orientation of the PhysicsRobot to the given position
+     * and angle. The robot will have its linear and angular velocity set to zero.
      *
      * @param position The new position of the robot
+     * @param angle The new angle of the robot
      */
-    void setPosition(const Point &position);
+    void setPosition(const Point &position, const Angle &angle);
```
Comment (jonathanlew) [inline src/software/simulation/physics/physics_robot.h]: this needs a new name

File: src/software/gui/standalone_simulator/widgets/standalone_simulator_draw_function_visualizer.h [hunk 5]
```diff
@@ -45,13 +46,19 @@ class StandaloneSimulatorDrawFunctionVisualizer : public DrawFunctionVisualizer
     // The robot currently being placed by the user, if any
     std::weak_ptr<PhysicsRobot> robot;
 
+    // Boolean to check if user shift clicked
+    bool shift_clicked = false;
```
Comment (jonathanlew) [inline src/software/gui/standalone_simulator/widgets/standalone_simulator_draw_function_visualizer.h]: this should be initialized in the constructor (also ctrl_clicked and r_clicked)

File: src/software/gui/standalone_simulator/widgets/standalone_simulator_draw_function_visualizer.h [hunk 6]
```diff
@@ -45,13 +46,19 @@ class StandaloneSimulatorDrawFunctionVisualizer : public DrawFunctionVisualizer
...
+
     // Boolean to check if the user ctrl clicked
     bool ctrl_clicked = false;
 
-    // The initial point when adding force to the ball, if any
+    // Boolean to check if user has r pressed
+    bool r_clicked = false;
+
+    // The final point when adding force to the ball, or rotating robot, if any
     Point initial_point;
```
Comment (jonathanlew) [inline src/software/gui/standalone_simulator/widgets/standalone_simulator_draw_function_visualizer.h]: maybe name this initial_click_point?

File: src/software/gui/standalone_simulator/widgets/standalone_simulator_draw_function_visualizer.cpp [hunk 7]
```diff
@@ -55,6 +69,16 @@ void StandaloneSimulatorDrawFunctionVisualizer::mouseReleaseEvent(QMouseEvent* e
...
+    else if (shift_clicked)
+    {
+        shift_clicked = false;
+    }
+    else if (r_clicked)
+    {
+        initial_point = Point(0, 0);
+        final_point   = Point(0, 0);
+        r_clicked     = false;
+    }
```
Comment (jonathanlew) [inline src/software/gui/standalone_simulator/widgets/standalone_simulator_draw_function_visualizer.cpp]: What if we removed the if/else around setting *_clicked to false? Similarly with initial_point and final_point

**Review by jonathanlew — CHANGES_REQUESTED** [review]
I noticed that the robot jumps to under the robot cursor. Perhaps initial_point needs to be set to the robot position

**Comment by PhilD71** [comment]
Made some changes! I added a function `resetFlags` which resets all the booleans to false, and resets the points to (0,0). This should simplify the if/else cases in `mouseReleaseEvent`. I fixed the robot jumping, cleaned up the fields in the `SSdrawFunctionVisualiser` and gave a more descriptive name to the function that sets position and orientation in PhysicsRobot.

File: src/software/gui/standalone_simulator/widgets/standalone_simulator_draw_function_visualizer.cpp [hunk 8]
```diff
@@ -13,6 +13,9 @@ StandaloneSimulatorDrawFunctionVisualizer::StandaloneSimulatorDrawFunctionVisual
 {
     // Let mouseMoveEvents be triggered even if a mouse button is not pressed
     this->setMouseTracking(true);
+
+    shift_clicked = false;
+    ctrl_clicked  = false;
```
Comment (jonathanlew) [inline src/software/gui/standalone_simulator/widgets/standalone_simulator_draw_function_visualizer.cpp]: ideally, these should set in the initializer list, i.e. `DrawFunctionVisualizer(parent), shift_clicked(false), ctrl_clicked(false)`

File: src/software/gui/standalone_simulator/widgets/standalone_simulator_draw_function_visualizer.cpp [hunk 9]
```diff
@@ -64,23 +91,48 @@ void StandaloneSimulatorDrawFunctionVisualizer::mouseMoveEvent(QMouseEvent* even
...
+            drawBallVelocity(scene, initial_click_point,
+                             initial_click_point - final_click_point,
+                             ball_speed_slow_color, ball_speed_fast_color);
+        };
+        return WorldDrawFunction(draw_function);
+    }
+    else
+    {
+        auto draw_function = [this](QGraphicsScene* scene) {
+            drawBallVelocity(scene, initial_click_point, Vector(0, 0),
```
Comment (jonathanlew) [inline src/software/gui/standalone_simulator/widgets/standalone_simulator_draw_function_visualizer.cpp]: we need some comments on what these two cases are about

File: src/software/gui/standalone_simulator/widgets/standalone_simulator_draw_function_visualizer.cpp [hunk 10]
```diff
@@ -21,21 +24,46 @@ void StandaloneSimulatorDrawFunctionVisualizer::setStandaloneSimulator(
...
+        shift_clicked        = true;
         Point point_in_scene = createPoint(mapToScene(event->pos()));
         robot                = standalone_simulator->getRobotAtPosition(point_in_scene);
     }
+    else if (r_clicked && standalone_simulator)
+    {
+        initial_click_point = createPoint(mapToScene(event->pos()));
+        robot = standalone_simulator->getRobotAtPosition(initial_click_point);
+        auto physics_robot  = robot.lock();
+        initial_click_point = physics_robot->position();
```
Comment (jonathanlew) [inline src/software/gui/standalone_simulator/widgets/standalone_simulator_draw_function_visualizer.cpp]: I get a segfault here when physics_robot is null. If we're going to do it like this, then we need to check if physics_robot is not null

File: src/software/gui/standalone_simulator/widgets/standalone_simulator_draw_function_visualizer.cpp [hunk 11]
```diff
@@ -64,23 +91,48 @@ void StandaloneSimulatorDrawFunctionVisualizer::mouseMoveEvent(QMouseEvent* even
...
+            physics_robot->setPositionAndOrientation(point_in_scene,
+                                                     physics_robot->orientation());
+        }
+        else if (r_clicked)
+        {
+            final_click_point = createPoint(mapToScene(event->pos()));
+            Vector vec        = Vector(final_click_point.x() - initial_click_point.x(),
+                                final_click_point.y() - initial_click_point.y());
+            Angle angle       = vec.orientation();
+            physics_robot->setPositionAndOrientation(initial_click_point, angle);
```
Comment (jonathanlew) [inline src/software/gui/standalone_simulator/widgets/standalone_simulator_draw_function_visualizer.cpp]: an alternative would be
```suggestion
            physics_robot->setPositionAndOrientation(physics_robot->position(), angle);
```

**Review by jonathanlew — CHANGES_REQUESTED** [review]
just a few more comments

File: src/software/gui/standalone_simulator/widgets/standalone_simulator_draw_function_visualizer.cpp (hunk 8 above)
Comment (PhilD71) [reply src/software/gui/standalone_simulator/widgets/standalone_simulator_draw_function_visualizer.cpp]: Did not know that about cpp! Done

File: src/software/gui/standalone_simulator/widgets/standalone_simulator_draw_function_visualizer.cpp (hunk 10 above)
Comment (PhilD71) [reply src/software/gui/standalone_simulator/widgets/standalone_simulator_draw_function_visualizer.cpp]: Yeah that was silly, I should have tested it more. I've fixed it and messed around with it trying to break it. Seems to work fine now

**Review by jonathanlew — APPROVED** [review]
LGTM! :rotating_light:

**Review by akhilveeraghanta — APPROVED** [review]
lgtm :fire:

**Pull request merged**